stdlib/public/SDK/Foundation/NSStringAPI.swift

```diff
@@ -35,30 +35,82 @@ func _toNSRange(_ r: Range<String.Index>) -> NSRange {
 }
 
 func _countFormatSpecifiers(_ a: String) -> Int {
-  // The implementation takes advantage of the fact that internal
-  // representation of String is UTF-16.  Because we only care about the ASCII
-  // percent character, we don't need to decode UTF-16.
-
+  // The implementation takes advantage of the fact that the internal
+  // representation of String is UTF-16.  Because we only care about ASCII
+  // characters (percent, dollar, zero, and nine), we don't need to decode UTF-16.
+  
+  enum ScanState {
+    case NoPercent
+    case SeenPercent
+    case ScanningDigits
+  }
+  
   let percentUTF16  = UTF16.CodeUnit(("%" as UnicodeScalar).value)
-  let notPercentUTF16: UTF16.CodeUnit = 0
-  var lastChar = notPercentUTF16 // anything other than % would work here
+  let dollarUTF16   = UTF16.CodeUnit(("$" as UnicodeScalar).value)
+  let zeroUTF16   = UTF16.CodeUnit(("0" as UnicodeScalar).value)
+  let nineUTF16   = UTF16.CodeUnit(("9" as UnicodeScalar).value)
   var count = 0
-
+  var maxPosValue = 0
+  var posValue = 0
+  var state = ScanState.NoPercent
+  var posMode = false
+  
   for c in a.utf16 {
-    if lastChar == percentUTF16 {
+    switch state {
+    case .NoPercent:
+      if c == percentUTF16 {
+        state = .SeenPercent
+      }
+      
+    case .SeenPercent:
       if c == percentUTF16 {
         // a "%" following this one should not be taken as literal
-        lastChar = notPercentUTF16
+        state = .NoPercent
+      }
+      else if c >= zeroUTF16 && c <= nineUTF16 {
+        posValue = Int(c) - Int(zeroUTF16)
+        state = .ScanningDigits
+      }
+      else {
+        if posMode {
+          // error condition: non-positional format seen after positional format(s)
```

Comment:
Could you please add `FIXME`s here and later, to add a proper error handling? And maybe file a bug, so that it's tracked.